src/HJSON/HJSONParser.php

```diff
@@ -308,6 +308,7 @@ private function error($m)
 
         $i = $this->at;
         while ($i > 0) {
+            $ch = mb_substr(mb_strcut($this->text, $i - 1), 0, 1);
             $i -= strlen($ch);
```

Comment:
I only confused about this.

It looks like some places using the `mb_*` functions and some places don't use.

I think it should use the `mb_*` functions for all source codes because of consistency and it will avoid some issues.

Reply:
The only reason I use `strlen` in that area is because `$i` and `$this->at` are counting in bytes instead of characters. Most of the `mbstring` library functions (except for `mb_strcut`) count in characters. It is generally faster and more efficient to use byte indexes instead of character indexes, so I compromise by letting `$ch` represent a utf8 character but increment `$i` by the number of bytes in `$ch`.

The reason `mbstring` implementations can be less efficient is because utf8 characters can vary in length, so `mbstring` has to start from the beginning of the string and search until it finds its desired character(s). However, ideally normal string functions can index directly into the middle of a string since bytes are of constant length.

For example, [`mb_substr`](http://php.net/manual/en/function.mb-substr.php#117764) has to loop through the entire string from the beginning until it finds the character index you are looking for. However, I think the normal `substr` function is more efficient than that. Feel free to correct me if I've mistaken.

I think eventually doing a smart split like qbolec said on the [`mb_substr`](http://php.net/manual/en/function.mb-substr.php#117764) page would help a lot with efficiency, but that would require a major overhaul of most of the parser code.

Reply:
Hi @dqsully, thank you for your reply.

It makes sense.

Some works need to use the `strlen` to calculate the bytes and others need to use `mb_strlen` to calculate characters.

Final part that you mention is about using the `mb_substr` to be more efficient than `substr`, but how many works should be changed on this pacakge?

I think the final part can have the another issue and it's not related to this issue currently.